### PR TITLE
Allow some of the special abilities to target non-players

### PR DESCRIPTION
**Pull request by Coolthulhu**
Updated some creature attacks to target generic Creatures, not just g->u.

Added uncanny_dodge to Creature. Reversed order of arguments on monster::sees( int&, Creature\* );
Removed unused argument from player::uncanny_dodge. is_u is now checked in the method.

Message handling gets kinda bloated, so I'll probably write something like add_msg_if_player, except generalized to Creature (add_msg_player_npc_monster), possibly with faction detection to make coloring automatic (m_bad for player and allies, m_neutral for others).

### CODE REVIEW

File: src/monattack.cpp [hunk 1]
```diff
@@ -165,7 +182,7 @@ void mattack::acid(monster *z, int index)
 
     int t;
     int junk = 0;
-    if( !z->sees( t, target ) ||
+    if( !z->sees( target, t ) ||
```
Comment (KA101) [inline src/monattack.cpp]: I'm sure there's a good reason for these but the code as written seemed to work.  What's up?

File: src/monattack.cpp [hunk 2]
```diff
@@ -1083,54 +1110,61 @@ void mattack::fungus_inject(monster *z, int index)
...
-    if (dam > 0) {
+    if( foe != nullptr && dam > 0 ) {
         //~ 1$s is monster name, 2$s bodypart in accusative
-        add_msg(m_bad, _("The %1$s sinks several needlelike barbs into your %2$s!"), z->name().c_str(),
-                body_part_name_accusative(hit).c_str());
+        foe->add_msg_if_player( m_bad, _("The %1$s sinks several needlelike barbs into your %2$s!"), z->name().c_str(),
+                                body_part_name_accusative(hit).c_str());
 
         if(one_in(15 - dam)) {
             g->u.add_effect("fungus", 200, num_bp, true);
```
Comment (KA101) [inline src/monattack.cpp]: Looks like that gives *you* a fungal infection if it hits anyone?

File: src/monattack.cpp [hunk 3]
```diff
@@ -3118,8 +3157,10 @@ void mattack::upgrade(monster *z, int index)
 {
     std::vector<int> targets;
     for (size_t i = 0; i < g->num_zombies(); i++) {
-        if (g->zombie(i).type->id == "mon_zombie" &&
-            rl_dist( z->pos(), g->zombie(i).pos() ) <= 5) {
+        monster &zed = g->zombie(i);
+        if( zed.type->id == "mon_zombie" &&
+            rl_dist( z->pos(), zed.pos() ) <= 5 &&
+            z->attitude_to( zed ) != Creature::Attitude::A_HOSTILE ) {
```
Comment (KA101) [inline src/monattack.cpp]: OK, so if I read this right it won't upgrade zeds that aren't on its side (yours or Team Zombie, depending).  Good catch.

File: src/monattack.cpp (hunk 1 above)
Comment (Coolthulhu) [inline src/monattack.cpp]: I wrote the monster::sees basing it on sees_player just below. sees_player had an optional argument, which had to be last argument and so had argument order different from all other "sees" function. monster::sees has no optional argument, so I changed it to look like all other "sees" function.

**Comment by Coolthulhu** [comment]
Anything else I missed?

File: src/monattack.cpp [hunk 4]
```diff
@@ -434,45 +453,39 @@ void mattack::science(monster *z, int index) // I said SCIENCE again!
...
-        add_msg(m_bad, _("The %s opens it's mouth and a beam shoots towards you!"),
-                z->name().c_str());
+        add_msg( msg_type, _("The %s opens it's mouth and a beam shoots towards %s!"),
+                z->name().c_str(), target->disp_name().c_str() );
         z->moves -= 400;
-        if (!g->u.uncanny_dodge()) {
-            if (g->u.get_dodge() > rng(0, 16) && !one_in(g->u.get_dodge())) {
+        if( !foe->uncanny_dodge() ) {
+            if( foe->get_dodge() > rng(0, 16) && !one_in( foe->get_dodge() ) ) {
                 add_msg(_("You dodge the beam!"));
```
Comment (BevapDin) [inline src/monattack.cpp]: This should probably be `foe->add_msg_...`

File: src/monattack.cpp [hunk 5]
```diff
@@ -697,31 +711,39 @@ void mattack::grow_vine(monster *z, int index)
...
+                    player *foe = dynamic_cast< player* >( critter );
                     body_part bphit = random_body_part();
                     //~ 1$s monster name(vine), 2$s bodypart in accusative
-                    add_msg(m_bad, _("The %1$s lashes your %2$s!"), z->name().c_str(),
-                            body_part_name_accusative(bphit).c_str());
+                    if( critter == &g->u ) {
+                        add_msg( m_bad, _("The %1$s lashes your %2$s!"), z->name().c_str(),
+                                 body_part_name_accusative(bphit).c_str() );
+                    } else if( foe != nullptr ) {
+                        add_msg( _("The %1$s lashes %s's %2$s!"), z->name().c_str(),
```
Comment (BevapDin) [inline src/monattack.cpp]: This block might need either a sees check, or use `add_msg_if_npc`, that function will ignore the message if the creature is not visible to the player. Same for the else branch.

File: src/monattack.cpp [hunk 6]
```diff
@@ -810,12 +835,12 @@ void mattack::spit_sap(monster *z, int index)
...
         return;
     }
-    if (g->u.uncanny_dodge() ) {
+    if( target->uncanny_dodge() ) {
         return;
     }
-    add_msg(m_bad, _("A glob of sap hits you!"));
-    g->u.deal_damage( z, bp_torso, damage_instance( DT_BASH, dam ) );
-    g->u.add_effect("sap", dam);
+    add_msg( msg_type, _("A glob of sap hits %s!"), target->disp_name().c_str() );
```
Comment (BevapDin) [inline src/monattack.cpp]: `target->add_msg_player_or_npc` would allow to keep the "a sap hits you!" part and includes the can_see for the targe.

**Comment by KA101** [comment]
Thanks.  Was pretty confident that the mutation beam wouldn't end well v. NPCs.

File: src/monattack.cpp [hunk 7]
```diff
@@ -434,79 +453,88 @@ void mattack::science(monster *z, int index) // I said SCIENCE again!
...
-        }
+    case 1: // Shock the target
+        // Just reuse the taze - it's a bit different (shocks torso vs all), but let's go for consistency here
+        taze( z, target );
         break;
     case 2: // Radioactive beam
-        add_msg(m_bad, _("The %s opens it's mouth and a beam shoots towards you!"),
-                z->name().c_str());
+        if( seen ) {
+            add_msg( msg_type, _("The %s opens it's mouth and a beam shoots towards %s!"),
```
Comment (KA101) [inline src/monattack.cpp]: Don't let the Slayer happen to you!
https://github.com/CleverRaven/Cataclysm-DDA/pull/10286

**Comment by Coolthulhu** [comment]
I'm pretty sure mut beam it doesn't need much to start working (there isn't much static mutation code and has no g->u where generic target should be), it's just that all mutation code uses add_msg rather than add_msg_if_player. Though NPCs might get head asploding from bodypart muts (trying to put scarf over beak etc.).

Can you fix the apostrophe on your side? Switching branches makes my entire objwin directory invalidated and full recompilation takes an hour on my toaster.

**Comment by KA101** [comment]
Apostrophe: sure.

NPCs: Ugh.  I inherited mutation.cpp and have been glad to get it to cooperate when I've needed to add infrastructure.

**Comment by KA101** [comment]
Oops, just ran into problems with the mouth beam in testing the sightline PR.  Segfaulting on the foe->uncanny_dodge line.  Stack trace: http://pastebin.com/kM3wDrZK

**Comment by Coolthulhu** [comment]
#10816 should fix the segfault.

**Pull request merged**